src/redist/targets/BundledTemplates.targets

```diff
@@ -1,55 +1,125 @@
+      <NupkgPathRelativeToPackageRoot>%(Identity)/%(PackageVersion)/%(Identity).%(PackageVersion).nupkg</NupkgPathRelativeToPackageRoot>
+      <RestoredNupkgPath>$(NuGetPackageRoot)$([MSBuild]::ValueOrDefault('%(NupkgPathRelativeToPackageRoot)', '').ToLower())</RestoredNupkgPath>
+      <Version>[%(PackageVersion)]</Version>
+    </Bundled30Templates>
+  </ItemGroup>
+  <ItemGroup>
+    <Bundled22Templates Update="@(Bundled22Templates)">
+      <NupkgPathRelativeToPackageRoot>%(Identity)/%(PackageVersion)/%(Identity).%(PackageVersion).nupkg</NupkgPathRelativeToPackageRoot>
+      <RestoredNupkgPath>$(NuGetPackageRoot)$([MSBuild]::ValueOrDefault('%(NupkgPathRelativeToPackageRoot)', '').ToLower())</RestoredNupkgPath>
+      <Version>[%(PackageVersion)]</Version>
+    </Bundled22Templates>
+  </ItemGroup>
+  <ItemGroup>
+    <Bundled21Templates Update="@(Bundled21Templates)">
+      <NupkgPathRelativeToPackageRoot>%(Identity)/%(PackageVersion)/%(Identity).%(PackageVersion).nupkg</NupkgPathRelativeToPackageRoot>
+      <RestoredNupkgPath>$(NuGetPackageRoot)$([MSBuild]::ValueOrDefault('%(NupkgPathRelativeToPackageRoot)', '').ToLower())</RestoredNupkgPath>
+      <Version>[%(PackageVersion)]</Version>
+    </Bundled21Templates>
   </ItemGroup>
 
-  <!-- Restore bundled templates via PackageReference -->
   <ItemGroup>
-    <PackageReference Include="@(BundledTemplate)" />
+    <PackageDownload Include="@(Bundled22Templates);@(Bundled21Templates)" />
   </ItemGroup>
 
+  <!-- Restore bundled templates via PackageReference -->
   <ItemGroup>
-    <BundledTemplate Update="@(BundledTemplate)">
-      <NupkgPathRelativeToPackageRoot>%(Identity)/%(Version)/%(Identity).%(Version).nupkg</NupkgPathRelativeToPackageRoot>
-      <RestoredNupkgPath>$(NuGetPackageRoot)$([MSBuild]::ValueOrDefault('%(NupkgPathRelativeToPackageRoot)', '').ToLower())</RestoredNupkgPath>
-    </BundledTemplate>
+    <PackageReference Include="@(Bundled30Templates)" />
   </ItemGroup>
 
   <Target Name="LayoutTemplates"
-          DependsOnTargets="SetupBundledComponents;CalculateTemplatesVersion">
+        DependsOnTargets="LayoutTemplatesForSDK;LayoutTemplatesFor30MSI;LayoutTemplatesFor22MSI;LayoutTemplatesFor21MSI" />
+    
+  <Target Name="LayoutTemplatesForSDK"
+          DependsOnTargets="SetupBundledComponents;CalculateTemplatesVersions">
     <ItemGroup Condition="!$(ProductMonikerRid.StartsWith('win'))">
       <BundledTemplate Remove="Microsoft.Dotnet.Wpf.ProjectTemplates" />
       <BundledTemplate Remove="Microsoft.Dotnet.WinForms.ProjectTemplates" />
     </ItemGroup>
-    <Copy SourceFiles="%(BundledTemplate.RestoredNupkgPath)"
-          DestinationFolder="$(RedistLayoutPath)templates/$(BundledTemplatesVersion)"/>
+    <Copy SourceFiles="%(Bundled30Templates.RestoredNupkgPath)"
+          DestinationFolder="$(RedistLayoutPath)templates/$(BundledTemplates30Version)"/>
+  </Target>
+
+  <Target Name="LayoutTemplatesFor30MSI"
```

Comment:
Not a big fan of all this repetition as well but not sure it is worth to try to remove it by operating on items here.

@dsplaisted what do you think?

Reply:
I think you'd probably need to build a subproject in order to avoid the repetition for the file copy tasks here.  I think I'm OK with leaving this as it is here.